sigma-dev/sigma
# Service Labels and Selectors

Closes #351 

### changelog 

- [x] model.go / API -- plugins now only have requirements, not optional. services have a set of string labels, and requirements are still a list of strings (but are now selectors instead of a list of explicit required services). services have no optional services. finally, pluginreqs now have the `min` field, specifying the minimum cardinality of that required service to get the other service to run. 
- [x] API convert 
- [x] validation for selectors and labels -- ensuring strings are of the right format
- [x] manager logic -- resolving selectors to explicit lists of services + checking cardinality 
- [x] updated `examples/` [with four new plugins that use different Mins](https://github.com/sigma-dev/sigma/tree/labels-selectors/examples/selectors)
- [x] update existing examples to use the new `requirements` format. 
- [x] [Docs PR](https://github.com/sigma-dev/sigma/pull/397)

### how to test 

- go into `./selectors/examples` and run `./load.sh` which sets up a bunch of services using labels + selectors. Can watch them run in the expected order, etc. 
- you can customize the services in that folder to experiment with more complex labels and selectors 
- you can also update any of the other examples that use requirements (goodbye-world, my-weather, coffee-process) to use labels. Right now the selectors in those services are very simple, ie. `service=hello-world-svc`.
